### DISCUSSION TITLE
how to make mypy happy about Json and not see it as str only

### DISCUSSION THREAD

**Discussion by yifan** [pydantic/pydantic#2660] [Question]
Hi,

I have a field in BaseModel as following:
```python
class X(BaseModel):
    record:Json[Dict[str,str]]
```
I am expecting to use `X.record` as dict in my code. The code would work, but **mypy** would complain because when doing type checking, `Json` is always seen as `str` according to the link below:

pydantic/pydantic/types.py

Line 712 in ae13be5
```python
 if TYPE_CHECKING: 
  json = str
```
How should I use Json and make mypy happy? Thank you.

**Comment by PrettyWood** [comment]
Hi @yifan 
Why not just write
```py
class X(BaseModel):
    record: Dict[str,str]
```
?
Especially when you say
> I am expecting to use `X.record` as dict in my code

**Reply by yifan** [reply]
I want to be able to use Json to parse the input json string, so that I could do:
```python
  X(record='{"key":"value"}')
```
It probably won't work as expect, I should use `json.loads` and keep things simple.

**Reply by PrettyWood** [reply]
Think about you "inner" model representation. Here you want to have a `dict`.
If you want to support extra types of inputs, you can just use validators to deserialize them.
Here I would probably do
```py
import json

from pydantic import BaseModel, validator


class X(BaseModel):
    record: dict[str, str]

    @validator('record', pre=True)
    def parse_if_not_dict(cls, v):
        if not isinstance(v, dict):
            return json.loads(v)
        return v
```

**Reply by yifan** [reply]
Thank you. What I am doing is to adapt BaseSettings to allow parsing from command line. Basically I am modifying BaseSetting value by scanning .__fields__ and command line arguments. 
```python
        parser = ArgumentParser(add_help=True)
        for name, field in self.__fields__.items():
            name = (self.Config.env_prefix + name).replace("_", "-")
            parser.add_argument(
                f"--{name}",
                type=field.type_,
                action=SetSettingsAction,
                help=field.field_info.title,
            )
        parser.parse_known_args(args)
```

I won't be able to add specific validator. I think I will need to use a different action to deal with field with type_ for dict.

**Reply by yifan** [reply]
Thank you. I think I have made the problem more complicated than it should be. I will use a str and do parsing when the value being used. It is not a major problem since it will be evaluated at setting up stage of the application anyway.  It is over engineering by myself.
